socketlabs.js

```diff
@@ -0,0 +1,106 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, you can obtain one at https://mozilla.org/MPL/2.0/.
+
+const EVENTS = {
+  VALIDATION: 'Validation',
+  COMPLAINT: 'Complaint',
+  FAILED: 'Failed',
+  DELIVERED: 'Delivered'
+}
+
+function marshallEvent (event) {
+  if (! event || event.Type === EVENTS.VALIDATION || ! event.Address || ! event.DateTime || ! event.Type) {
+    return
+  }
+
+  const timestamp = new Date(event.DateTime).toISOString()
+  const mail = marshallMailObject(event, timestamp)
+
+  switch (event.Type) {
+    case EVENTS.FAILED:
+      return { mail, ...marshallBounceEvent(event, timestamp) }
+
+    case EVENTS.DELIVERED:
+      return { mail, ...marshallDeliveryEvent(event, timestamp) }
+
+    case EVENTS.COMPLAINT:
+      return { mail, ...marshallComplaintEvent(event, timestamp) }
+  }
+}
+
+function marshallMailObject (event, timestamp) {
+  return {
+    timestamp,
+    messageId: event.MessageId
```

Comment:
Fwiw, `MessageId` is only useful if we had it earlier too, when the message was sent. Otherwise it won't match the key we set in Redis and the queues process won't pull out any metadata.

IIRC, you mentioned that SocketLabs doesn't return us the message id when we send the email? If that's still the case, we might need to think about some alternative approach, like including our own generated message id as a header when we send the email, and then reading that header here (if we have access to them here).

Reply:
> IIRC, you mentioned that SocketLabs doesn't return us the message id when we send the email?

Hmm, hadn't thought about that, but it's right

Reply:
> I'll check if we can get the headers from the email messge, that would be nice.

Apparently we can't... http://www.socketlabs.com/api-reference/notification-api/

Reply:
@brizental, just looking at the Socketlabs, it looks like we can set the `MessageId` ourselves in the injection API:

https://www.socketlabs.com/api-reference/injection-api/

Then we'd get that same value out here. Or am I misunderstanding?

Reply:
@philbooth Ah! That's true! Alright, that makes everything much easier. Thanks, Phil 👍